target-component-accessor-compiler/ksp/src/main/kotlin/me/tatarka/inject/target/component/accessor/compiler/ksp/TargetComponentAccessorProcessor.kt

```diff
@@ -0,0 +1,146 @@
+import com.google.devtools.ksp.symbol.KSFunctionDeclaration
+import com.google.devtools.ksp.symbol.KSTypeReference
+import com.google.devtools.ksp.symbol.Modifier
+import com.squareup.kotlinpoet.CodeBlock
+import com.squareup.kotlinpoet.FileSpec
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.KModifier
+import com.squareup.kotlinpoet.ksp.toKModifier
+import com.squareup.kotlinpoet.ksp.toTypeName
+import com.squareup.kotlinpoet.ksp.writeTo
+import me.tatarka.inject.annotations.TargetComponentAccessor
+
+class TargetComponentAccessorProcessor(
+    private val codeGenerator: CodeGenerator,
+    private val logger: KSPLogger,
+) : SymbolProcessor {
+    @Suppress("LoopWithTooManyJumpStatements")
+    override fun process(resolver: Resolver): List<KSAnnotated> {
+        val symbols = resolver.getSymbolsWithAnnotation(
+            requireNotNull(TargetComponentAccessor::class.qualifiedName),
+        )
+
+        for (symbol in symbols) {
+            if (symbol !is KSFunctionDeclaration) {
+                logger.error("TargetComponentAccessor should be applied to a function")
+                continue
+            }
+
+            val funcName = symbol.simpleName.asString()
+
+            // the generated actual funs will be annotated with TargetComponentAccessor
+            // KSP will process them as well so we need to ignore them
+            if (Modifier.ACTUAL in symbol.modifiers && symbol.findExpects().firstOrNull() != null) {
+                continue
+            }
+
+            if (Modifier.EXPECT !in symbol.modifiers) {
+                logger.error("$funcName should be an expect fun")
+                continue
+            }
+
+            val funcReceiver = symbol.validateReceiver()
+            if (funcReceiver == null) {
+                logger.error("$funcName should be declared as an extension of a @Component's companion")
+                continue
+            }
+
+            val funcReturn = symbol.validateReturnType()
+            if (funcReturn == null) {
+                logger.error("$funcName's return type should be a type annotated with @Component")
+                continue
+            }
+
+            generate(
+                symbol = symbol,
+                name = funcName,
+                receiverType = funcReceiver,
+                returnType = funcReturn,
+                codeGenerator = codeGenerator
+            )
+        }
+
+        return emptyList()
+    }
+}
+
+private fun KSFunctionDeclaration.validateReceiver(): KSTypeReference? {
+    val funcReceiver = extensionReceiver
+    val funcReceiverName = funcReceiver?.resolve()?.declaration?.qualifiedName?.asString()
+    return funcReceiver?.takeIf {
+        funcReceiverName != null && funcReceiverName.endsWith("Companion")
+    }
+}
+
+private fun KSFunctionDeclaration.validateReturnType(): KSDeclaration? {
+    return returnType?.resolve()?.declaration?.takeIf { funcReturn ->
```

Comment:
If the component itself is generated (say with some kind of anvil-like extension) then this may not resolve in the current round. You should check this and delay processing (see: https://kotlinlang.org/docs/ksp-multi-round.html)